File: packages/sdk-python/src/melaya/pipelines.py
"""Pipelines API — overview, runs, traces, cron schedules, and pipeline lifecycle.

Maps to ``/api/v1/private/overview/pipeline*``, ``/api/v1/private/runs/:runId/traces``,
``/api/v1/private/pipeline-schedule``, and ``/api/v1/private/pipelines/*``.

Example
-------
>>> from melaya import Melaya
>>> m = Melaya(api_key="mk_...")
>>> runs = m.pipelines.list(project="my-project", limit=20)
>>> traces = m.pipelines.traces(run_id="run-123")
>>> m.pipelines.upsert_schedule("my-project", "nightly-report", cron="0 2 * * *")
>>> cfg = m.pipelines.create(name="my-pipe", project="my-project")
>>> result = m.pipelines.run("my-pipe", project="my-project")
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional, TypedDict
from urllib.parse import quote

from .platform_types import JsonDict


class TraceSummary(TypedDict, total=False):
    traceId: str
    traceName: str
    startTime: str
    endTime: str
    status: Any
    spanCount: int
    totalTokens: int


class TracesPage(TypedDict):
    """Paginated envelope returned by ``GET /api/v1/private/runs/:runId/traces``."""
    list: List[TraceSummary]
    total: int
    page: int
    pageSize: int


class DeleteTracesResult(TypedDict, total=False):
    """Return shape for DELETE /api/v1/private/runs/:runId/traces."""
    deletedSpans: int
    requestedTraces: int


class PipelinesAPI:
    def __init__(self, request: Any) -> None:
        self._request = request

    # ── Overview ────────────────────────────────────────────────────────────────

    def overview(self) -> JsonDict:
        """Dashboard overview: usage stats, active strategies, recent runs."""
        return self._request("GET", "/api/v1/private/overview")

    def model_prices(self) -> JsonDict:
        """Get pricing data for available AI models."""
        return self._request("GET", "/api/v1/private/overview/model-prices")

    def chart_data(self) -> JsonDict:
        """Get chart data for overview dashboard (cost/usage over time)."""
        return self._request("GET", "/api/v1/private/overview/chart")

    def cost_breakdown(self) -> JsonDict:
        """Get cost breakdown by model/provider."""
        return self._request("GET", "/api/v1/private/overview/cost-breakdown")

    def count(self) -> JsonDict:
        """Get count of pipeline runs grouped by status."""
        return self._request("GET", "/api/v1/private/overview/pipeline-count")

    def list(
        self,
        *,
        project: Optional[str] = None,
        pipeline_name: Optional[str] = None,
        status: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        page: Optional[int] = None,
    ) -> List[JsonDict]:
        """Get paginated list of pipeline runs."""
        params: Dict[str, Any] = {}
        if project is not None:
            params["project"] = project
        if pipeline_name is not None:
            params["pipelineName"] = pipeline_name
        if status is not None:
            params["status"] = status
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        if page is not None:
            params["page"] = page
        return self._request("GET", "/api/v1/private/overview/pipelines", params=params)

    def recent(self) -> List[JsonDict]:
        """Get most recent pipeline runs for dashboard widget."""
        return self._request("GET", "/api/v1/private/overview/pipelines/recent")

    # ── Traces ──────────────────────────────────────────────────────────────────

    def traces(self, run_id: str) -> TracesPage:
        """List traces for a run.

        Returns a paginated envelope ``{"list": [...], "total": int, "page": int, "pageSize": int}``.
        The envelope is nested under a ``"data"`` key in the raw API response;
        this method unwraps it so callers receive the inner page dict directly.
        """
        result = self._request("GET", f"/api/v1/private/runs/{quote(run_id, safe='')}/traces")
        # Server returns { data: { list, total, page, pageSize } }
        if isinstance(result, dict) and "data" in result:
            return result["data"]
        return result

    def trace(self, run_id: str, trace_id: str) -> JsonDict:
        """Get a single trace by ID."""
        return self._request(
            "GET",
            f"/api/v1/private/runs/{quote(run_id, safe='')}/traces/{quote(trace_id, safe='')}",
        )

    def trace_stats(self, run_id: str, trace_id: str) -> JsonDict:
        """Get statistics for a specific trace."""
        return self._request(
            "GET",
            f"/api/v1/private/runs/{quote(run_id, safe='')}/traces/{quote(trace_id, safe='')}/stats",
        )

    def delete_traces(self, run_id: str) -> DeleteTracesResult:
        """Delete all traces for a run by runId.

        Sends ``DELETE /api/v1/private/runs/:runId/traces`` with no request body.

        Returns
        -------
        DeleteTracesResult
            ``{"deletedSpans": int}`` (``requestedTraces`` is included when available).
        """
        return self._request("DELETE", f"/api/v1/private/runs/{quote(run_id, safe='')}/traces")

    # ── Schedule ────────────────────────────────────────────────────────────────

    def list_schedules(self) -> List[JsonDict]:
        """List all pipeline schedules accessible to the caller."""
        return self._request("GET", "/api/v1/private/pipeline-schedule")

    def get_schedule(self, project: str, pipeline_name: str) -> JsonDict:
        """Get schedule status for a pipeline."""
        return self._request(
            "GET",
            f"/api/v1/private/pipeline-schedule/{quote(project, safe='')}/{quote(pipeline_name, safe='')}",
        )

    def upsert_schedule(
        self,
        project: str,
        pipeline_name: str,
        *,
        cron: str,
        config: Optional[Dict[str, Any]] = None,
    ) -> JsonDict:
        """Create or update a pipeline schedule (cron expression + optional pipeline config)."""
        body: Dict[str, Any] = {"cron": cron}
        if config is not None:
            body["config"] = config
        return self._request(
            "PUT",
            f"/api/v1/private/pipeline-schedule/{quote(project, safe='')}/{quote(pipeline_name, safe='')}",
            json=body,
        )

    def pause_schedule(self, project: str, pipeline_name: str) -> JsonDict:
        """Pause a pipeline schedule."""
        return self._request(
            "POST",
            f"/api/v1/private/pipeline-schedule/{quote(project, safe='')}/{quote(pipeline_name, safe='')}/pause",
        )

    def resume_schedule(self, project: str, pipeline_name: str) -> JsonDict:
        """Resume a paused pipeline schedule."""
        return self._request(
            "POST",
            f"/api/v1/private/pipeline-schedule/{quote(project, safe='')}/{quote(pipeline_name, safe='')}/resume",
        )

    # ── Pipeline lifecycle ───────────────────────────────────────────────────────

    def list_pipelines(self) -> List[JsonDict]:
        """List all pipeline configurations accessible to the caller.

        Returns
        -------
        list of pipeline config dicts under the ``pipelines`` key.
        """
        result = self._request("GET", "/api/v1/private/pipelines")
        if isinstance(result, dict) and "pipelines" in result:
            return result["pipelines"]
        return result

    def create(
        self,
        *,
        name: str,
        project: str,
        description: Optional[str] = None,
        **config: Any,
    ) -> JsonDict:
        """Create a new pipeline configuration.

        Parameters
        ----------
        name:
            Unique pipeline name within the project.
        project:
            Project slug the pipeline belongs to.
        description:
            Optional human-readable description.
        **config:
            Additional pipeline configuration fields forwarded to the API.
        """
        body: Dict[str, Any] = {"name": name, "project": project, **config}
        if description is not None:
            body["description"] = description
        return self._request("POST", "/api/v1/private/pipelines", json=body)

    def get(self, name: str, *, project: Optional[str] = None) -> JsonDict:
        """Fetch a pipeline configuration by name.

        Parameters
        ----------
        name:
            Pipeline name (URL-encoded automatically).
        project:
            Optional project filter query parameter.
        """
        params: Dict[str, Any] = {}
        if project is not None:
            params["project"] = project
        return self._request("GET", f"/api/v1/private/pipelines/{quote(name, safe='')}", params=params)

    def update(self, name: str, *, config: Dict[str, Any], project: str) -> JsonDict:
        """Update an existing pipeline configuration.

        Parameters
        ----------
        name:
            Pipeline name to update (URL-encoded automatically).
        config:
            New pipeline configuration dict to persist.
        project:
            Project the pipeline belongs to.
        """
        body: Dict[str, Any] = {"config": config, "project": project}
        return self._request("PUT", f"/api/v1/private/pipelines/{quote(name, safe='')}", json=body)

    def remove(self, name: str, *, project: str) -> JsonDict:
        """Delete a pipeline configuration.

        Parameters
        ----------
        name:
            Pipeline name to delete (URL-encoded automatically).
        project:
            Project the pipeline belongs to (sent as query parameter).
        """
        return self._request(
            "DELETE",
            f"/api/v1/private/pipelines/{quote(name, safe='')}",
            params={"project": project},
        )

    def run(
        self,
        name: str,
        *,
        project: Optional[str] = None,
        execution_target: Optional[str] = None,
        studio_url: Optional[str] = None,
        env_overrides: Optional[Dict[str, str]] = None,
    ) -> JsonDict:
        """Trigger a pipeline run.

        Parameters
        ----------
        name:
            Pipeline name to run (URL-encoded automatically).
        project:
            Optional project override.
        execution_target:
            Optional execution target override: local-runner or cloud-spawn.
        studio_url:
            Optional Studio URL override.
        env_overrides:
            Optional environment variable overrides for this run.

        Returns
        -------
        dict with ``run_id`` and ``queued`` keys.
        """
        body: Dict[str, Any] = {}
        if project is not None:
            body["project"] = project
        if execution_target is not None:
            body["executionTarget"] = execution_target
        if studio_url is not None:
            body["studio_url"] = studio_url
        if env_overrides is not None:
            body["env_overrides"] = env_overrides
        return self._request("POST", f"/api/v1/private/pipelines/{quote(name, safe='')}/run", json=body)

    def run_ids(self, name: str) -> List[str]:
        """List all run IDs for a pipeline.

        Parameters
        ----------
        name:
            Pipeline name (URL-encoded automatically).

        Returns
        -------
        list of run ID strings under the ``run_ids`` key.
        """
        result = self._request("GET", f"/api/v1/private/pipelines/{quote(name, safe='')}/runs")
        if isinstance(result, dict) and "run_ids" in result:
            return result["run_ids"]
        return result

    def run_status(self, name: str, run_id: str) -> JsonDict:
        """Get the status of a specific pipeline run.

        Parameters
        ----------
        name:
            Pipeline name (URL-encoded automatically).
        run_id:
            Run ID (URL-encoded automatically).

        Returns
        -------
        dict with ``runId``, ``status``, ``createdAt``, ``executionTarget``, and ``cost``.
        """
        return self._request(
            "GET",
            f"/api/v1/private/pipelines/{quote(name, safe='')}/runs/{quote(run_id, safe='')}",
        )

    def cancel_run(self, name: str, run_id: str) -> JsonDict:
        """Cancel a running or queued pipeline run.

        Parameters
        ----------
        name:
            Pipeline name (URL-encoded automatically).
        run_id:
            Run ID to cancel (URL-encoded automatically).
        """
        return self._request(
            "DELETE",
            f"/api/v1/private/pipelines/{quote(name, safe='')}/runs/{quote(run_id, safe='')}",
        )

    def outputs(self, name: str) -> List[JsonDict]:
        """List all output artifacts produced by a pipeline.

        Parameters
        ----------
        name:
            Pipeline name (URL-encoded automatically).
        """
        return self._request("GET", f"/api/v1/private/pipelines/{quote(name, safe='')}/outputs")

    def output(self, name: str, path: str, *, download: bool = False) -> JsonDict:
        """Fetch a single output artifact by path.

        Each segment of ``path`` is URL-encoded individually and joined with
        ``/`` so that slashes in individual path components are preserved as
        separators rather than escaped.

        Parameters
        ----------
        name:
            Pipeline name (URL-encoded automatically).
        path:
            Artifact path (e.g. ``"results/report.json"``). Each ``/``-delimited
            segment is percent-encoded independently.
        download:
            When ``True``, appends ``?download=1`` to request a download URL.
        """
        encoded_path = "/".join(quote(segment, safe="") for segment in path.split("/"))
        params: Dict[str, Any] = {}
        if download:
            params["download"] = 1
        return self._request(
            "GET",
            f"/api/v1/private/pipelines/{quote(name, safe='')}/outputs/{encoded_path}",
            params=params,
        )

    def preview_code(self, config: Dict[str, Any]) -> JsonDict:
        """Generate a code preview for a pipeline configuration without saving it.

        Parameters
        ----------
        config:
            Pipeline configuration dict to preview.
        """
        return self._request("POST", "/api/v1/private/pipelines/preview-code", json=config)

    def tools(self) -> JsonDict:
        """Get the tools registry available to pipelines."""
        return self._request("GET", "/api/v1/private/pipelines/tools")

    def subagents(self) -> JsonDict:
        """Get the subagents registry available to pipelines."""
        return self._request("GET", "/api/v1/private/pipelines/subagents")

    def instantiate_template(
        self,
        template_id: str,
        *,
        name: str,
        project: str,
        overrides: Optional[Dict[str, Any]] = None,
    ) -> JsonDict:
        """Instantiate a pipeline from a template.

        Parameters
        ----------
        template_id:
            Template ID to instantiate (URL-encoded automatically).
        name:
            Name for the new pipeline.
        project:
            Project to create the pipeline in.
        overrides:
            Optional configuration overrides applied on top of the template defaults.

        Returns
        -------
        dict with a ``pipeline`` key containing the new pipeline config.
        """
        body: Dict[str, Any] = {"name": name, "project": project}
        if overrides is not None:
            body["overrides"] = overrides
        return self._request(
            "POST",
            f"/api/v1/private/templates/{quote(template_id, safe='')}/instantiate",
            json=body,
        )

    def build_with_ai(self, brief: Dict[str, Any]) -> JsonDict:
        """Build a pipeline configuration from a natural language brief using AI.

        Parameters
        ----------
        brief:
            Dict describing the desired pipeline behaviour. Shape is open-ended
            and passed directly to the AI build endpoint.

        Returns
        -------
        Generated pipeline configuration dict.
        """
        return self._request("POST", "/api/v1/private/ai/build-pipeline/sync", json=brief)

File: packages/sdk-rust/src/pipelines.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;

use crate::client::HttpClient;
use crate::error::Result;

// ── Trace types ───────────────────────────────────────────────────────────────

/// A single trace summary row returned by `GET /runs/:runId/traces`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TraceSummary {
    pub trace_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trace_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end_time: Option<String>,
    /// Integer status code: 0 = ok, 1 = unset/warn, 2 = error (OTel convention).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub span_count: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_tokens: Option<u64>,
}

/// Paginated envelope returned by `getTraces` / `GET /runs/:runId/traces`.
/// Matches `{ data: { list, total, page, pageSize } }` from the server.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TracesPage {
    pub list: Vec<TraceSummary>,
    pub total: u64,
    pub page: u64,
    pub page_size: u64,
}

/// Response from `DELETE /runs/:runId/traces`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteTracesResult {
    pub deleted_spans: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub requested_traces: Option<u64>,
}

// ── Lifecycle structs ─────────────────────────────────────────────────────────

/// Options for running a pipeline.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineRunOptions {
    /// Project the pipeline belongs to.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project: Option<String>,
    /// Where to execute: local-runner or cloud-spawn.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_target: Option<String>,
    /// Studio URL used for cloud executions (injected by the Studio).
    #[serde(rename = "studio_url", skip_serializing_if = "Option::is_none")]
    pub studio_url: Option<String>,
    /// Per-run environment variable overrides.
    #[serde(rename = "env_overrides", skip_serializing_if = "Option::is_none")]
    pub env_overrides: Option<Value>,
}

/// Response returned by the pipeline run endpoint.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineRunAccepted {
    /// Unique identifier for the queued run.
    #[serde(rename = "run_id")]
    pub run_id: String,
    /// Whether the run was queued successfully.
    pub queued: bool,
}

/// Live status of a pipeline run.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineRunStatus {
    /// Unique identifier for the run.
    pub run_id: String,
    /// Current status string (e.g. `"queued"`, `"running"`, `"success"`, `"failed"`).
    pub status: String,
    /// ISO-8601 creation timestamp.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_at: Option<String>,
    /// Where the run executed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_target: Option<String>,
    /// Structured run cost totals (if available).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost: Option<Value>,
}

/// Percent-encode a single URL path segment.
///
/// Encodes every byte that is not an RFC 3986 unreserved character
/// (`ALPHA / DIGIT / - . _ ~`). This is safe for embedding in any path
/// segment where `/` must remain a path delimiter.
fn encode_segment(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'.' | b'_' | b'~' => {
                out.push(b as char);
            }
            _ => {
                out.push('%');
                out.push(
                    char::from_digit((b >> 4) as u32, 16)
                        .unwrap()
                        .to_ascii_uppercase(),
                );
                out.push(
                    char::from_digit((b & 0xF) as u32, 16)
                        .unwrap()
                        .to_ascii_uppercase(),
                );
            }
        }
    }
    out
}

/// Pipelines API — overview, runs, traces, and cron schedules.
#[derive(Clone)]
pub struct PipelinesAPI {
    http: HttpClient,
}

impl PipelinesAPI {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    // ── Overview ────────────────────────────────────────────────────────────

    /// Dashboard overview: usage stats, active strategies, recent runs.
    pub async fn overview(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.get("/api/v1/private/overview", &q).await
    }

    /// Get pricing data for available AI models.
    pub async fn model_prices(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/overview/model-prices", &q)
            .await
    }

    /// Get chart data for the overview dashboard (cost/usage over time).
    pub async fn chart_data(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.get("/api/v1/private/overview/chart", &q).await
    }

    /// Get cost breakdown by model/provider.
    pub async fn cost_breakdown(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/overview/cost-breakdown", &q)
            .await
    }

    /// Count of pipeline runs grouped by status.
    pub async fn count(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/overview/pipeline-count", &q)
            .await
    }

    /// Paginated list of pipeline runs.
    pub async fn list(
        &self,
        project: Option<&str>,
        pipeline_name: Option<&str>,
        status: Option<&str>,
        limit: Option<u32>,
        offset: Option<u32>,
        page: Option<u32>,
    ) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("project", project.map(str::to_owned));
        q.insert("pipelineName", pipeline_name.map(str::to_owned));
        q.insert("status", status.map(str::to_owned));
        q.insert("limit", limit.map(|v| v.to_string()));
        q.insert("offset", offset.map(|v| v.to_string()));
        q.insert("page", page.map(|v| v.to_string()));
        self.http
            .get("/api/v1/private/overview/pipelines", &q)
            .await
    }

    /// Most recent pipeline runs for a dashboard widget.
    pub async fn recent(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/overview/pipelines/recent", &q)
            .await
    }

    // ── Traces ──────────────────────────────────────────────────────────────

    /// List all traces for a run (paginated).
    ///
    /// Returns a `TracesPage` envelope: `{ list, total, page, pageSize }`.
    pub async fn traces(&self, run_id: &str) -> Result<TracesPage> {
        let q = HashMap::new();
        let enc_run = encode_segment(run_id);
        let raw = self
            .http
            .get(&format!("/api/v1/private/runs/{enc_run}/traces"), &q)
            .await?;
        // Server wraps in { data: { list, total, page, pageSize } }
        let inner = raw.get("data").cloned().unwrap_or(raw);
        Ok(serde_json::from_value(inner)?)
    }

    /// Get a single trace by ID.
    pub async fn trace(&self, run_id: &str, trace_id: &str) -> Result<Value> {
        let q = HashMap::new();
        let enc_run = encode_segment(run_id);
        let enc_trace = encode_segment(trace_id);
        self.http
            .get(
                &format!("/api/v1/private/runs/{enc_run}/traces/{enc_trace}"),
                &q,
            )
            .await
    }

    /// Get statistics for a specific trace.
    pub async fn trace_stats(&self, run_id: &str, trace_id: &str) -> Result<Value> {
        let q = HashMap::new();
        let enc_run = encode_segment(run_id);
        let enc_trace = encode_segment(trace_id);
        self.http
            .get(
                &format!("/api/v1/private/runs/{enc_run}/traces/{enc_trace}/stats"),
                &q,
            )
            .await
    }

    /// Delete all traces for a run by run ID.
    ///
    /// Calls `DELETE /api/v1/private/runs/:runId/traces` (no body).
    /// Returns `{ deletedSpans, requestedTraces? }`.
    pub async fn delete_traces(&self, run_id: &str) -> Result<DeleteTracesResult> {
        let q = HashMap::new();
        let enc_run = encode_segment(run_id);
        let raw = self
            .http
            .delete(&format!("/api/v1/private/runs/{enc_run}/traces"), &q)
            .await?;
        Ok(serde_json::from_value(raw)?)
    }

    // ── Schedule ────────────────────────────────────────────────────────────

    /// List all pipeline schedules accessible to the caller.
    pub async fn list_schedules(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.get("/api/v1/private/pipeline-schedule", &q).await
    }

    /// Get the schedule for a specific pipeline.
    pub async fn get_schedule(&self, project: &str, pipeline_name: &str) -> Result<Value> {
        let q = HashMap::new();
        let enc_proj = encode_segment(project);
        let enc_name = encode_segment(pipeline_name);
        self.http
            .get(
                &format!("/api/v1/private/pipeline-schedule/{enc_proj}/{enc_name}"),
                &q,
            )
            .await
    }

    /// Create or update a pipeline schedule (cron expression + optional config).
    pub async fn upsert_schedule(
        &self,
        project: &str,
        pipeline_name: &str,
        cron: &str,
        config: Option<&Value>,
    ) -> Result<Value> {
        let mut body = json!({ "cron": cron });
        if let Some(cfg) = config {
            body["config"] = cfg.clone();
        }
        let enc_proj = encode_segment(project);
        let enc_name = encode_segment(pipeline_name);
        self.http
            .put(
                &format!("/api/v1/private/pipeline-schedule/{enc_proj}/{enc_name}"),
                &body,
            )
            .await
    }

    /// Pause a pipeline schedule.
    pub async fn pause_schedule(&self, project: &str, pipeline_name: &str) -> Result<Value> {
        let enc_proj = encode_segment(project);
        let enc_name = encode_segment(pipeline_name);
        self.http
            .post(
                &format!("/api/v1/private/pipeline-schedule/{enc_proj}/{enc_name}/pause"),
                &json!({}),
            )
            .await
    }

    /// Resume a paused pipeline schedule.
    pub async fn resume_schedule(&self, project: &str, pipeline_name: &str) -> Result<Value> {
        let enc_proj = encode_segment(project);
        let enc_name = encode_segment(pipeline_name);
        self.http
            .post(
                &format!("/api/v1/private/pipeline-schedule/{enc_proj}/{enc_name}/resume"),
                &json!({}),
            )
            .await
    }

    // ── Pipeline lifecycle (CRUD + run management) ───────────────────────────

    /// List all pipelines accessible to the caller.
    ///
    /// Returns `{ pipelines: [...] }`.
    pub async fn list_pipelines(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.get("/api/v1/private/pipelines", &q).await
    }

    /// Create a new pipeline.
    ///
    /// `config` may contain any additional pipeline configuration fields
    /// beyond `name`, `project`, and `description`.
    pub async fn create(
        &self,
        name: &str,
        project: &str,
        description: Option<&str>,
        config: Option<&Value>,
    ) -> Result<Value> {
        let mut body = json!({
            "name": name,
            "project": project,
        });
        if let Some(d) = description {
            body["description"] = json!(d);
        }
        if let Some(cfg) = config {
            if let Some(obj) = cfg.as_object() {
                for (k, v) in obj {
                    body[k] = v.clone();
                }
            }
        }
        self.http.post("/api/v1/private/pipelines", &body).await
    }

    /// Get a pipeline by name, optionally scoped to a project.
    pub async fn get(&self, name: &str, project: Option<&str>) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("project", project.map(str::to_owned));
        let encoded = encode_segment(name);
        self.http
            .get(&format!("/api/v1/private/pipelines/{encoded}"), &q)
            .await
    }

    /// Update a pipeline's configuration.
    ///
    /// `config` is the full pipeline config object; `project` scopes the lookup.
    pub async fn update(&self, name: &str, config: &Value, project: &str) -> Result<Value> {
        let body = json!({ "config": config, "project": project });
        let encoded = encode_segment(name);
        self.http
            .put(&format!("/api/v1/private/pipelines/{encoded}"), &body)
            .await
    }

    /// Delete a pipeline.
    pub async fn delete_pipeline(&self, name: &str, project: &str) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("project", Some(project.to_owned()));
        let encoded = encode_segment(name);
        self.http
            .delete(&format!("/api/v1/private/pipelines/{encoded}"), &q)
            .await
    }

    /// Enqueue a pipeline run and return the run ID.
    pub async fn run(
        &self,
        name: &str,
        opts: Option<PipelineRunOptions>,
    ) -> Result<PipelineRunAccepted> {
        let body = opts
            .map(|o| serde_json::to_value(o).unwrap_or(json!({})))
            .unwrap_or(json!({}));
        let encoded = encode_segment(name);
        let raw = self
            .http
            .post(&format!("/api/v1/private/pipelines/{encoded}/run"), &body)
            .await?;
        Ok(serde_json::from_value(raw)?)
    }

    /// List all run IDs for a pipeline.
    ///
    /// Returns `{ run_ids: [...] }`.
    pub async fn run_ids(&self, name: &str) -> Result<Value> {
        let q = HashMap::new();
        let encoded = encode_segment(name);
        self.http
            .get(&format!("/api/v1/private/pipelines/{encoded}/runs"), &q)
            .await
    }

    /// Get the status of a specific run.
    pub async fn run_status(&self, name: &str, run_id: &str) -> Result<PipelineRunStatus> {
        let q = HashMap::new();
        let enc_name = encode_segment(name);
        let enc_run = encode_segment(run_id);
        let raw = self
            .http
            .get(
                &format!("/api/v1/private/pipelines/{enc_name}/runs/{enc_run}"),
                &q,
            )
            .await?;
        Ok(serde_json::from_value(raw)?)
    }

    /// Cancel a running or queued pipeline run.
    pub async fn cancel_run(&self, name: &str, run_id: &str) -> Result<Value> {
        let q = HashMap::new();
        let enc_name = encode_segment(name);
        let enc_run = encode_segment(run_id);
        self.http
            .delete(
                &format!("/api/v1/private/pipelines/{enc_name}/runs/{enc_run}"),
                &q,
            )
            .await
    }

    /// List all output artifacts for a pipeline.
    pub async fn outputs(&self, name: &str) -> Result<Value> {
        let q = HashMap::new();
        let encoded = encode_segment(name);
        self.http
            .get(&format!("/api/v1/private/pipelines/{encoded}/outputs"), &q)
            .await
    }

    /// Get a specific output artifact by path.
    ///
    /// `output_path` will be split on `/` and each segment individually
    /// percent-encoded before joining, so forward slashes in segment names
    /// are preserved as path separators.
    ///
    /// Pass `download: true` to append `?download=1` (returns a download URL).
    pub async fn output(&self, name: &str, output_path: &str, download: bool) -> Result<Value> {
        let encoded_name = encode_segment(name);
        let encoded_path: String = output_path
            .split('/')
            .map(encode_segment)
            .collect::<Vec<_>>()
            .join("/");
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        if download {
            q.insert("download", Some("1".to_owned()));
        }
        self.http
            .get(
                &format!("/api/v1/private/pipelines/{encoded_name}/outputs/{encoded_path}"),
                &q,
            )
            .await
    }

    /// Preview the generated code for a pipeline config without persisting it.
    pub async fn preview_code(&self, config: &Value) -> Result<Value> {
        self.http
            .post("/api/v1/private/pipelines/preview-code", config)
            .await
    }

    /// List available agent tools in the platform registry.
    pub async fn tools(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.get("/api/v1/private/pipelines/tools", &q).await
    }

    /// List available subagents in the platform registry.
    pub async fn subagents(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get("/api/v1/private/pipelines/subagents", &q)
            .await
    }

    /// Instantiate a pipeline from a template.
    ///
    /// `overrides` is an optional JSON object of config overrides applied on
    /// top of the template defaults.
    pub async fn instantiate_template(
        &self,
        template_id: &str,
        name: &str,
        project: &str,
        overrides: Option<&Value>,
    ) -> Result<Value> {
        let mut body = json!({ "name": name, "project": project });
        if let Some(ov) = overrides {
            body["overrides"] = ov.clone();
        }
        let enc_id = encode_segment(template_id);
        self.http
            .post(
                &format!("/api/v1/private/templates/{enc_id}/instantiate"),
                &body,
            )
            .await
    }

    /// Generate a pipeline configuration from a natural-language brief using AI.
    ///
    /// `brief` is a free-form `Value` (string or structured object) describing
    /// the pipeline to build. Returns the generated pipeline config.
    pub async fn build_with_ai(&self, brief: &Value) -> Result<Value> {
        self.http
            .post("/api/v1/private/ai/build-pipeline/sync", brief)
            .await
    }

    /// Return the current public server version.
    pub async fn server_version(&self) -> Result<Value> {
        let query = HashMap::new();
        self.http.get("/api/v1/version", &query).await
    }
}
